### PR TITLE
Change MPI Communication for Large Data with CELEBA Dataset

### PR DESCRIPTION
**Pull request by yim0331**
- Add Celeba example.
- Change MPI communication to handle large data
- Allocate clients to multiple GPUs in multi GPUs - multi Process environment

### CODE REVIEW

File: src/appfl/config/config.py [hunk 1]
```diff
@@ -70,6 +70,8 @@ class Config:
     logginginfo: DictConfig = OmegaConf.create({})
     summary_file: str = ""
 
+    # GPU configuration
+    num_gpu: int = 1
```
Comment (kibaekkim) [inline src/appfl/config/config.py:74]: Is this parameter used for the runs with mpi? Can this be used with gRPC?

File: src/appfl/run_mpi.py [hunk 2]
```diff
@@ -245,26 +240,36 @@ def run_client(
...
-            outfile[cid],
-            test_dataloader,
-            **cfg.fed.args,
+    clientpergpu = math.ceil(num_clients/cfg.num_gpu)    
+    clients = []
+    for _, cid in enumerate(num_client_groups[comm_rank - 1]):
+        ## We assume to have as many GPUs as the number of MPI processes.
+        if "cuda" in cfg.device:
+            gpuindex = int(math.floor(cid/clientpergpu))            
+            device = f"cuda:{gpuindex}"
```
Comment (kibaekkim) [inline src/appfl/run_mpi.py:249]: I think `gpuindex` may better be a configuration parameter that can be set for each client. For example, we may better set this parameter in the example file.

File: src/appfl/config/config.py (hunk 1 above)
Comment (yim0331) [reply src/appfl/config/config.py:74]: Yes, this parameter is independent of the MPI processes. The parameter is used only if the CUDA is available according to the torch.cuda.is_available() function. If the CUDA is available, the default config is set to 1, and replaced by torch.cuda.device_count() function.

File: src/appfl/run_mpi.py (hunk 2 above)
Comment (yim0331) [reply src/appfl/run_mpi.py:249]: I'm not sure if I understand correctly your comment. Does it mean that the user should set what GPU to assign for each client or set the number of clients for each GPU? 
I believe it would be nice to have a feature that distributes the clients to all available GPUs to utilize all the resources in the machine.

**Review by kibaekkim — COMMENTED** [review]
It is not clearly how `num_gpu` is used. In `femnist.py` example, it sets the number of GPUs available (per node?) on cluster. Whenever we have more than one GPU, it calls `torch.nn.DataParallel`. What does it do?

File: src/appfl/run_mpi.py [hunk 3]
```diff
@@ -75,10 +78,15 @@ def run_server(
     else:
         cfg.validation = False
 
+    # Broadcast limitation according to the procs numbers
+    intmax = 2100000000 # INT_MAX,2147483647, does not fit on the MPI send buffer size
+    recvlimit = math.floor(intmax/(comm_size-1)) # The total recv size should be less than intmax due to the MPI recv buffer size
```
Comment (Zilinghan) [inline src/appfl/run_mpi.py:83]: Out of curiosity, is there any document for this? How this number 2100000000 is obtained, and why recvlimit is intmax/commsize-1

File: examples/femnist.py [hunk 4]
```diff
@@ -132,6 +132,9 @@ def main():
 
     args.num_clients = len(train_datasets)
     model = get_model(args)
+    cfg.num_gpu = args.num_gpu
+    if cfg.num_gpu > 1:
+        model = torch.nn.DataParallel(model)
```
Comment (Zilinghan) [inline examples/femnist.py:137]: I wonder why we use DataParallel here? From its [document](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html), it says: It is recommended to use [DistributedDataParallel](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel), instead of this class, to do multi-GPU training, even if there is only a single node. See: [Use nn.parallel.DistributedDataParallel instead of multiprocessing or nn.DataParallel](https://pytorch.org/docs/stable/notes/cuda.html#cuda-nn-ddp-instead) and [Distributed Data Parallel](https://pytorch.org/docs/stable/notes/ddp.html#ddp).

File: src/appfl/run_mpi.py [hunk 5]
```diff
@@ -244,47 +251,104 @@ def run_client(
...
-                pin_memory=True,
-            ),
-            cfg,
-            outfile[cid],
-            test_dataloader,
-            **cfg.fed.args,
+    clientpergpu = math.ceil(num_clients/cfg.num_gpu)    
+    clients = []
+    for _, cid in enumerate(num_client_groups[comm_rank - 1]):
+        ## We assume to have as many GPUs as the number of MPI processes.
```
Comment (Zilinghan) [inline src/appfl/run_mpi.py:257]: Does this line of comment still hold?

File: examples/celeba.py [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+
+dir = os.getcwd() + "/datasets/RawData/%s" % (args.dataset)
+
+def get_data(comm: MPI.Comm):
+    # test data for a server 
+    test_data_raw = {}
+    test_data_input = []
+    test_data_label = []
+    imagedir = dir + "/raw/img_align_celeba/"
```
Comment (Zilinghan) [inline examples/celeba.py:64]: I think we should as shell script for downloading the dataset, or instructions on how to download it. Also, we need to update the ReadME.md file in the examples folder

**Review by Zilinghan — CHANGES_REQUESTED** [review]

File: examples/femnist.py (hunk 4 above)
Comment (yim0331) [reply examples/femnist.py:137]: @Zilinghan The DataParallel is the legacy code for testing the Torch model-level parallelism. As APPFL assigns the model on the single-process, and single thread in a multi-process environment with MPI, APPFL does not need to implement the DataParallel and DistributedDataParallel.

File: src/appfl/run_mpi.py (hunk 5 above)
Comment (yim0331) [reply src/appfl/run_mpi.py:257]: Yes, the assumption is still held for the benchmarking performance (probably the best performance by [this reference](https://saturncloud.io/blog/how-to-share-model-and-session-between-processes-in-inference-with-nvidia-gpu/#:~:text=GPUs%20have%20limited%20memory%20and,decreased%20performance%20and%20even%20crashes.)). However, the updated code also allows users to run a non-matching combination of the GPU and MPI process to run and utilize the GPU as much as possible.

**Comment by yim0331** [comment]
@kibaekkim 
The set num_gpu is used to utilize the multiple GPUs if there is more than one available GPUs. The configuration is used to assign the models to different GPUs while constructing the models in [run_mpi.py](https://github.com/APPFL/APPFL/blob/a2e3b06c4285395378db097e1e1ef9bfb5fc23a7/src/appfl/run_mpi.py#L254-L263)
Now, I think it's better to move the check the number of available GPUs in run_mpi just before constructing the models, so APPFL can always utilize all the available GPUs. What do you think?

The DataParallel is the legacy code for testing the Torch model-level parallelism. As APPFL assigns the model on the single-process, and single thread in a multi-process environment with MPI, APPFL does not need to implement the DataParallel. I will remove it.

File: src/appfl/run_mpi.py (hunk 3 above)
Comment (kibaekkim) [reply src/appfl/run_mpi.py:83]: @Zilinghan good point!
@yim0331 : I think the size may be dependent to the integer type too: `int8`, `int16`, `int32`, ...

**Comment by kibaekkim** [comment]
> @kibaekkim The set num_gpu is used to utilize the multiple GPUs if there is more than one available GPUs. The configuration is used to assign the models to different GPUs while constructing the models in [run_mpi.py](https://github.com/APPFL/APPFL/blob/a2e3b06c4285395378db097e1e1ef9bfb5fc23a7/src/appfl/run_mpi.py#L254-L263) 

I am still not following you 100%. To my understanding, `num_gpu` is used to allocate each GPU to a group of "FL clients" before creating each client model. Is that correct?

> Now, I think it's better to move the check the number of available GPUs in run_mpi just before constructing the models, so APPFL can always utilize all the available GPUs. What do you think?

And what do you suggest here?

> The DataParallel is the legacy code for testing the Torch model-level parallelism. As APPFL assigns the model on the single-process, and single thread in a multi-process environment with MPI, APPFL does not need to implement the DataParallel. I will remove it.

@yim0331 : Just for clarification, I guess "legacy code" means the code **you** wrote and do not use anymore...? If it's not used, OK to remove it.

File: examples/femnist.py (hunk 4 above)
Comment (kibaekkim) [reply examples/femnist.py:137]: So will this be removed as mentioned below?

**Comment by yim0331** [comment]
@kibaekkim 

> I am still not following you 100%. To my understanding, num_gpu is used to allocate each GPU to a group of "FL clients" before creating each client model. Is that correct?
Yes it's correct

> And what do you suggest here?
Instead of checking the num_gpu on the example file(user define code), the variable can be checked just before allocating each GPU to a group of "FL clients" in [run_mpi.py](https://github.com/APPFL/APPFL/blob/a2e3b06c4285395378db097e1e1ef9bfb5fc23a7/src/appfl/run_mpi.py#L254-L263)

> Just for clarification, I guess "legacy code" means the code you wrote and do not use anymore...? If it's not used, OK to remove it.
Yes, it's not used. I will remove it.

**Comment by yim0331** [comment]
@kibaekkim @Zilinghan 
The updates since the last review requests are below:

- **Add CELEBA example document**: The data and pre-processing script references are added. The instruction to set the pre-processed file is added.
- **Remove legacy code**: The torch.nn.DataParallel code is removed.
- **Change the num_gpu implementation**: The checking code for the available GPUs is moved from femnist.py to run_mpi.py because the num_gpu variable is used for model creation with only the multi-GPU and multi-procs environment with MPI.
- **Change the MPI maximum buffer size assumption**: The size was changed from 2100000000 to (2^31) - (8 * comm_size) - 9. The equation is from the observation of the patterns of buffer overhead. However, the reason for the overhead is not clear, so [this issue](https://github.com/APPFL/APPFL/issues/140) is opened to address it. 
- **Resolve the conflict from the updated main**: Add the metric variable while creating the model.

**Comment by kibaekkim** [comment]
@yim0331 The PR looks good to me. Can you edit the title and description for this PR?

**Review by Zilinghan — APPROVED** [review]
The PR looks good to me
